pkg/trace/api/normalizer_test.go

```diff
@@ -33,277 +34,350 @@ func testSpan() *pb.Span {
 		assert.Equal(t, expName, s.Name)
-	}
-}
-
-func TestNormalizeNameFailure(t *testing.T) {
-	invalidNames := []string{
-		"",                                // Empty.
-		"/",                               // No alphanumerics.
-		"//",                              // Still no alphanumerics.
-		strings.Repeat("x", MaxNameLen+1), // Too long.
-	}
-	s := testSpan()
-	for _, v := range invalidNames {
-		s.Name = v
-		assert.Error(t, normalize(s))
+		assert.Equal(t, newTagStats(), ts)
 	}
 }
 
 func TestNormalizeResourcePassThru(t *testing.T) {
-	s := testSpan()
+	ts := newTagStats()
+	s := newTestSpan()
 	before := s.Resource
-	normalize(s)
+	assert.NoError(t, normalize(ts, s))
 	assert.Equal(t, before, s.Resource)
+	assert.Equal(t, newTagStats(), ts)
 }
 
 func TestNormalizeEmptyResource(t *testing.T) {
-	s := testSpan()
+	ts := newTagStats()
+	s := newTestSpan()
 	s.Resource = ""
-	assert.Error(t, normalize(s))
+	assert.NoError(t, normalize(ts, s))
+	assert.Equal(t, s.Resource, s.Name)
+	assert.Equal(t, tsMalformed(&info.SpansMalformed{ResourceEmpty: 1}), ts)
 }
 
 func TestNormalizeTraceIDPassThru(t *testing.T) {
-	s := testSpan()
+	ts := newTagStats()
+	s := newTestSpan()
 	before := s.TraceID
-	normalize(s)
+	assert.NoError(t, normalize(ts, s))
 	assert.Equal(t, before, s.TraceID)
+	assert.Equal(t, newTagStats(), ts)
 }
 
 func TestNormalizeNoTraceID(t *testing.T) {
-	s := testSpan()
+	ts := newTagStats()
+	s := newTestSpan()
 	s.TraceID = 0
-	normalize(s)
-	assert.NotEqual(t, 0, s.TraceID)
+	assert.Error(t, normalize(ts, s))
+	assert.Equal(t, tsDropped(&info.TracesDropped{TraceIDZero: 1}), ts)
 }
 
 func TestNormalizeSpanIDPassThru(t *testing.T) {
-	s := testSpan()
+	ts := newTagStats()
+	s := newTestSpan()
 	before := s.SpanID
-	normalize(s)
+	assert.NoError(t, normalize(ts, s))
 	assert.Equal(t, before, s.SpanID)
+	assert.Equal(t, newTagStats(), ts)
 }
 
 func TestNormalizeNoSpanID(t *testing.T) {
-	s := testSpan()
+	ts := newTagStats()
+	s := newTestSpan()
 	s.SpanID = 0
-	normalize(s)
-	assert.NotEqual(t, 0, s.SpanID)
+	assert.Error(t, normalize(ts, s))
+	assert.Equal(t, tsDropped(&info.TracesDropped{SpanIDZero: 1}), ts)
 }
 
 func TestNormalizeStartPassThru(t *testing.T) {
-	s := testSpan()
+	ts := newTagStats()
+	s := newTestSpan()
 	before := s.Start
-	normalize(s)
+	assert.NoError(t, normalize(ts, s))
 	assert.Equal(t, before, s.Start)
+	assert.Equal(t, newTagStats(), ts)
 }
 
 func TestNormalizeStartTooSmall(t *testing.T) {
-	s := testSpan()
+	ts := newTagStats()
+	s := newTestSpan()
 	s.Start = 42
-	assert.Error(t, normalize(s))
-}
-
-func TestNormalizeStartTooLarge(t *testing.T) {
-	s := testSpan()
-	s.Start = time.Now().Add(15 * time.Minute).UnixNano()
-	assert.Error(t, normalize(s))
+	before := s.Start
+	assert.NoError(t, normalize(ts, s))
+	assert.Equal(t, tsMalformed(&info.SpansMalformed{InvalidStartDate: 1}), ts)
+	assert.True(t, s.Start > before, "start should have been reset to current time")
```

Comment:
You can use https://golang.org/pkg/time/#Time.Before

Reply:
But then need to construct `time.Time` struct, where we're only working with `int64` here. Maybe simplest just to leave it like this?

Reply:
Yep, you're right. I missed that. It's fine to leave it like this. 👍